### PR TITLE
Added year-picker, by modifying a copy of the date-picker

### PR DESCRIPTION
**Pull request by joakimgk**
New attempt, this time done on the "next" branch. Checked that there are no lint errors, but I have not made any tests (in year-calendar/year-calendar.service.spec.ts). Hope this is a useable starting point, at least, to assess my contribution. The Year Picker works on the Demo page.

### CODE REVIEW

**Comment by vlio20** [comment]
@joakimgk, can you please post a link to the demo page?

**Comment by vlio20** [comment]
please note that you are also have errors in the Travis build

**Comment by joakimgk** [comment]
Well, I mean the demo in demo/demo/demo.component.html (as included in index.html when the project is run)

**Comment by joakimgk** [comment]
Yes, I noticed that. Could you please point me in the right direction? Are they complaining about missing tests?

**Comment by vlio20** [comment]
could you please create a github page? it's easy - because it's part of the build of this component.
you just need to run:
switch to release branch
npm run build:demo
git add -A
git commit -m "adding demo"
git subtree push --prefix dist origin gh-pages

**Comment by joakimgk** [comment]
Did it (from my "next" branch, which is where my modifications are). This creates a demo page on my GitHub?

**Comment by vlio20** [comment]
yes, I will take a look at the build as soon as I can

**Comment by joakimgk** [comment]
Thanks! So, where do you find this demo page, then? :)

**Comment by vlio20** [comment]
After posting it, it will be in your repo github pages page

**Comment by joakimgk** [comment]
Did you find the demo page?

**Comment by vlio20** [comment]
Nope

**Comment by joakimgk** [comment]
What does "switch to release branch" mean? Since all my modifications are on the "next" branch, I expect this is where I should create the demo. However, "npm run build:demo" yields no modifications, so "git add -A" and "git commit" does nothing.

**Comment by vlio20** [comment]
the problem that on the next branch the demo artefacts are ignored, switch to `release`  branch and merge to it the next branch. Then you should be able to run `build:demo` and then you will be able to add the demo pages as they are not ignored in this page.

**Comment by joakimgk** [comment]
I have tried to switch to the `release` branch, and merge in my changes from `next`, but there is just no way I am able to do the `push`. I've tried to make sure branches are up-to-date, but I still get errors when doing `git subtree push --prefix dist origin gh-pages`:

> error: failed to push some refs to 'https://github.com/joakimgk/angular-datepicker.git'
> hint: Updates were rejected because a pushed branch tip is behind its remote
> hint: counterpart. Check out this branch and integrate the remote changes
> hint: (e.g. 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Comment by vlio20** [comment]
You don't need to push, just the subtree command

**Comment by joakimgk** [comment]
Sorry, I don't understand. The error comes from the command `git subtree push --prefix dist origin gh-pages`, which involves a push, right? Anyway, while struggling to make a demo page, I've also been looking at the failed build. I fixed a typo in my `year-calendar.component.spec.ts`, and, after adding `"es6", "dom"` to the `tsconfig.spec.json` file, there are only two remaining failing tests -- but those are in files I have not modified: 

- `demo.component.spec.ts` invokes an empty constructor of `DemoComponent`, but `demo.component.ts` has no empty constructor. I fixed this (?) by simply passing in `null`:
```
describe('Component: Demo', () => {
  it('should create an instance', () => {
    const component = new DemoComponent(null);
    expect(component).toBeTruthy();
  });
});
```

- a similar test failed in `date-picker.directive.spec.ts` (line 20), in invoking the `DatePickerDirective` constructor. I had to add another `null`, to account for the fifth argument (even though it is decorated with `@Optional`).

**So, if `npm test -- -watch=false` yields SUCCESS (on my "next" branch), would you say it's OK for me to attempt another pull request?**

**Comment by vlio20** [comment]
Correct. From the error I see that you need to pull first.

Regarding tests. Please note that there are e2e tests as well. Also note that Master and Next are both passing the build.

**Comment by joakimgk** [comment]
Did it finally go through? :-)

**Comment by vlio20** [comment]
Not yet. I hope I will have time today or tomorrow

**Comment by vlio20** [comment]
Is there a demo already?

**Comment by joakimgk** [comment]
No, the demo thing I just had to give up. I would want to fix it, but I'd need some very specific instructions. My status is, I'm on the next branch, which is in sync with the same brach in your repo (right?). However, as I said before, when I switch to the release branch, and merge next into that, then `git add -A` and `commit` says no files are modified and branches are up to date. Yet, there are problems when I try the `git subtree` command.

**Comment by vlio20** [comment]
You may need to pull the release branch before.

**Comment by joakimgk** [comment]
Well, hopefully you can use this pull request to determine whether it's possible to merge this contribution into your project :-)

**Comment by vlio20** [comment]
I will clone your project and checkout the demo locally.

File: src/app/date-picker/date-picker-directive-config.model.ts [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import {TDrops, TOpens} from '../common/types/poistions.type';
 import {IDayCalendarConfig} from '../day-calendar/day-calendar-config.model';
 import {IMonthCalendarConfig} from '../month-calendar/month-calendar-config';
+import {IYearCalendarConfig} from '../year-calendar/year-calendar-config';
 import {ITimeSelectConfig} from '../time-select/time-select-config.model';
 
-export interface IDatePickerDirectiveConfig extends IDayCalendarConfig, IMonthCalendarConfig, ITimeSelectConfig {
+export interface IDatePickerDirectiveConfig extends IDayCalendarConfig, IMonthCalendarConfig, IYearCalendarConfig, ITimeSelectConfig {
```
Comment (vlio20) [inline src/app/date-picker/date-picker-directive-config.model.ts]: too long, please break to lines

File: src/app/demo/demo/demo.component.html [hunk 2]
```diff
@@ -292,6 +297,28 @@ <h2 class="dp-page-header">Angular Date/Time/Month Picker</h2>
...
       </div>
 
+      <div class="dp-picker-year" *ngIf="pickerMode === 'yearPicker'">
+        <dp-date-picker id="yearPicker"
+                        name="yearPicker"
+                        #datePicker
+                        #monthPicker="ngModel"
+                        [(ngModel)]="date"
+                        (ngModelChange)="log($event)"
+                        [mode]="'year'"
```
Comment (vlio20) [inline src/app/demo/demo/demo.component.html]: you can just `mode="year"`

File: src/app/year-calendar/year-calendar-config.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  isYearDisabledCallback?: (date: Moment) => boolean;
+  allowMultiSelect?: boolean;
+  yearFormat?: string;
+  yearFormatter?: (month: Moment) => string;
+  format?: string;
+  isNavHeaderBtnClickable?: boolean;
+  yearBtnFormat?: string;
+  yearBtnFormatter?: (day: Moment) => string;
+  yearBtnCssClassCallback?: (day: Moment) => string;
+  multipleYearsNavigateBy?: number;
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar-config.ts]: what will this do in a year picker?

File: src/app/year-calendar/year-calendar-config.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import {ICalendar, ICalendarInternal} from '../common/models/calendar.model';
+import {ECalendarValue} from '../common/types/calendar-value-enum';
+
+export interface IConfig {
+  isYearDisabledCallback?: (date: Moment) => boolean;
+  allowMultiSelect?: boolean;
+  yearFormat?: string;
+  yearFormatter?: (month: Moment) => string;
+  format?: string;
+  isNavHeaderBtnClickable?: boolean;
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar-config.ts]: what will this do?

File: src/app/year-calendar/year-calendar-config.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  yearBtnCssClassCallback?: (day: Moment) => string;
+  multipleYearsNavigateBy?: number;
+  showMultipleYearsNavigation?: boolean;
+  returnedValueType?: ECalendarValue;
+  showGoToCurrent?: boolean;
+}
+
+export interface IYearCalendarConfig extends IConfig,
+                                              ICalendar {
+
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar-config.ts]: redundant

File: src/app/year-calendar/year-calendar.component.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<div class="dp-year-calendar-container">
+  <dp-calendar-nav
+      [label]="navLabel"
+      [showLeftNav]="showLeftNav"
+      [showLeftSecondaryNav]="showSecondaryLeftNav"
+      [showRightNav]="showRightNav"
+      [showRightSecondaryNav]="showSecondaryRightNav"
+      [isLabelClickable]="componentConfig.isNavHeaderBtnClickable"
+      [showGoToCurrent]="shouldShowCurrent()"
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.html]: please don't use getter functions in template, use variables instead, this was implemented in 2.6.0 cross all components

File: src/app/year-calendar/year-calendar.component.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+})
+export class YearCalendarComponent implements OnInit, OnChanges, ControlValueAccessor, Validator {
+  @Input() config: IYearCalendarConfig;
+  @Input() displayDate: Moment;
+  @Input() minDate: Moment;
+  @Input() maxDate: Moment;
+  @HostBinding('class') @Input() theme: string;
+
+  @Output() onSelect: EventEmitter<IYear> = new EventEmitter();
+  @Output() onNavHeaderBtnClick: EventEmitter<null> = new EventEmitter();
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.ts]: should be void?

File: src/app/year-calendar/year-calendar.component.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    this.selected = this.utilsService
+      .updateSelected(this.componentConfig.allowMultiSelect, this.selected, year, 'year');
+    this.yearMatrix = this.yearCalendarService
+      .generateYear(this.componentConfig, this.currentDateView, this.selected);
+    this.onSelect.emit(year);
+  }
+
+  onLeftNav() {
+    this.currentDateView = this.currentDateView.clone().subtract(12, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.ts]: should be changed when currentDateView is changed, please see other components

File: src/app/year-calendar/year-calendar.component.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    if (isOutsideRange) {
+      navigateBy = this.currentDateView.year() - this.componentConfig.min.year();
+    }
+    this.currentDateView = this.currentDateView.clone().subtract(navigateBy, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);  // remove?
+  }
+
+  onRightNav() {
+    this.currentDateView = this.currentDateView.clone().add(12, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);  // remove?
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.ts]: remove???

File: src/app/year-calendar/year-calendar.component.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+  onRightSecondaryNav() {
+    let navigateBy = this.componentConfig.multipleYearsNavigateBy;
+    const isOutsideRange = this.componentConfig.max &&
+      this.componentConfig.max.year() - this.currentDateView.year() < navigateBy;
+
+    if (isOutsideRange) {
+      navigateBy = this.componentConfig.max.year() - this.currentDateView.year();
+    }
+    this.currentDateView = this.currentDateView.clone().add(navigateBy, 'year');
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.ts]: add empty line before this row

File: src/app/year-calendar/year-calendar.component.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);  // remove?
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.component.ts]: remove??

File: src/app/year-calendar/year-calendar.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import {YearCalendarService} from './year-calendar.service';
+import {IYear} from './year.model';
+
+describe('Service: YearCalendarService', () => {
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [YearCalendarService, UtilsService]
+    });
+  });
+
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.service.spec.ts]: many redundant empty lines

File: src/app/year-calendar/year-calendar.service.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        const _year = {
+          date,
+          selected: !!selected.find(s => index.isSame(s, 'year')),
+          currentYear: index.isSame(moment(), 'year'),
+          disabled: this.isYearDisabled(date, config),
+          text: this.getYearBtnText(config, date)
+        };
+
+        index.add(1, 'year');
+        return _year;
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.service.ts]: add line break before return

File: src/app/year-calendar/year-calendar.service.ts [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    return min ? min.isBefore(currentYearView, 'year') : true;
+  }
+
+  shouldShowRight(max: Moment, currentYearView: Moment): boolean {
+    return max ? max.isAfter(currentYearView, 'year') : true;
+  }
+
+  getHeaderLabel(config: IYearCalendarConfig, year: Moment): string {
+    return null;
+    /*
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.service.ts]: please remove if not required

File: src/app/year-calendar/year-calendar.service.ts [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    return year.format(config.yearFormat);
+    */
+  }
+
+  getYearBtnText(config: IYearCalendarConfig, year: Moment): string {
+    if (config.yearBtnFormatter) {
+      return config.yearBtnFormatter(year);
+    }
+    return year.format(config.yearBtnFormat);
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar.service.ts]: new line before return

File: src/app/year-calendar/year-calendar-config.ts (hunk 4 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: Right -- I'll omit this, the formatting parts. Year can only be year/YYYY, as I'm sure is your point.  The year calendar is molded from the month calendar, at first very closely to get it running. This detailed feedback from you is precisely what I was hoping for, to help me finish the code.

File: src/app/year-calendar/year-calendar-config.ts (hunk 5 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: OK

File: src/app/year-calendar/year-calendar.component.html (hunk 6 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.html]: But this is the same way in `month-calendar/month-calendar.component.html`, isn't it?

File: src/app/year-calendar/year-calendar.component.ts (hunk 8 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: Isn't this equivalent to the same method, in `month-calendar.component.ts`? I don't understand your comment.

File: src/app/year-calendar/year-calendar.component.ts (hunk 9 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: As above -- I've created the year calendar from the design of the month calendar. In `month-calendar.component.ts`, there is a, perhaps, strange difference between methods `onLeftNav()` and `onRightNav()`; only the former actually updates the `yearMonths` object. Why doesn't the "right" handler also update the object? I opted to update `yearMatrix` in both methods, but marked the code to draw your attention.

File: src/app/year-calendar/year-calendar.component.ts (hunk 10 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: OK

File: src/app/year-calendar/year-calendar.component.ts (hunk 11 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: -- see above --

File: src/app/year-calendar/year-calendar.service.spec.ts (hunk 12 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.service.spec.ts]: OK

File: src/app/year-calendar/year-calendar.service.ts (hunk 13 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.service.ts]: OK

File: src/app/year-calendar/year-calendar.service.ts (hunk 14 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.service.ts]: OK. You agree this is not needed? I considered making a header label of sorts, "2017-2028", but I'm not sure that's better than simply no header.

File: src/app/year-calendar/year-calendar.service.ts (hunk 15 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.service.ts]: OK

File: src/app/year-calendar/year-calendar.component.html (hunk 6 above)
Comment (vlio20) [reply src/app/year-calendar/year-calendar.component.html]: this might be something I missed, thanks for pointing out. Will fix on the month calendar.

File: src/app/year-calendar/year-calendar.component.ts (hunk 8 above)
Comment (vlio20) [reply src/app/year-calendar/year-calendar.component.ts]: take a look at https://github.com/vlio20/angular-datepicker/blob/master/src/app/month-calendar/month-calendar.component.ts#L86

**Comment by joakimgk** [comment]
Thanks for the very helpful feedback! Not sure why the build failed this time, to be honest, but I'll fix the things you point out and make a new push tomorrow. Hopefully it will build, then ;)

**Comment by vlio20** [comment]
Thanks for the PR. 
There are some parts that should be added:
1. e2e
2. documentation
3. Integration with month picker (same as day and month). There should be a way to jump from day to month to day (all should be available via configuration).

File: src/app/year-calendar/year-calendar-config.ts (hunk 4 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: s

Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: Looking at this now, and I'm not sure what, specifically, you're refering to. As far as I can see, the Year Calendar Config needs to specify a year format string, even if it can't be anything else than `YYYY`. However, it seems I can do without the `yearFormatter`. Can you please elaborate?

File: src/app/year-calendar/year-calendar-config.ts (hunk 3 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: Again, what are you refering to here? All the four lines, or any one in particular?

File: src/app/year-calendar/year-calendar-config.ts (hunk 5 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: ... however, this is identical to `month-calendar-config.ts`. I don't understand your comment, can I simply remove the `IYearCalendarConfig` interface?

File: src/app/year-calendar/year-calendar.component.ts (hunk 7 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: Please explain

File: src/app/year-calendar/year-calendar.component.ts (hunk 8 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: Yes, and my `currentDateView` method in `year-calendar.component.ts` does the same (please note I renamed the `this.yearMonths` to `this.yearMatrix`, for consistency): 
<snip>
    set currentDateView(current: Moment) {
        this._currentDateView = current.clone();
        this.yearMatrix = this.yearCalendarService
          .generateYear(this.componentConfig, this._currentDateView, this.selected);
</snip>

Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: (didn't find a multi-line code formatting)

File: src/app/year-calendar/year-calendar.component.ts (hunk 9 above)
Comment (joakimgk) [reply src/app/year-calendar/year-calendar.component.ts]: Any thoughts on this? :)

Comment (vlio20) [reply src/app/year-calendar/year-calendar.component.ts]: this is very odd... I need to check. Sorry for the long responses...

File: src/app/year-calendar/year-calendar-config.ts [hunk 16]
```diff
@@ -6,7 +6,6 @@ export interface IConfig {
   isYearDisabledCallback?: (date: Moment) => boolean;
   allowMultiSelect?: boolean;
   yearFormat?: string;
-  yearFormatter?: (month: Moment) => string;
```
Comment (vlio20) [inline src/app/year-calendar/year-calendar-config.ts]: why did you remove that?

Comment (joakimgk) [reply src/app/year-calendar/year-calendar-config.ts]: Ah, I think I need more specific instructions ;-) Hope you can point me in the right direction, especially wrt hooking the year calendar up to the month and date calendars. If the other things you point out are minor things, please help me get started :-)

**Comment by joakimgk** [comment]
Is there perhaps any technical documentation on how the datepicker is set up? I mean, how the different calendars work by themselves, and how they work together? I haven't had time to pick apart the existing code, and only have a working understanding (of the parts encountered while making the year calendar).

**Comment by vlio20** [comment]
There is a documentation for each component, but you will understand each configuration by its template

File: .gitignore [hunk 17]
```diff
@@ -36,3 +36,5 @@ testem.log
 Thumbs.db
 
 dist
+
+*.bak
```
Comment (vlio20) [inline .gitignore]: ?

File: src/app/date-picker/date-picker-config.model.ts [hunk 18]
```diff
@@ -20,12 +21,14 @@ export interface IConfig {
 export interface IDatePickerConfig extends IConfig,
                                            IDayCalendarConfig,
                                            IMonthCalendarConfig,
+                                           IYearCalendarConfig,
                                            ITimeSelectConfig {
```
Comment (vlio20) [inline src/app/date-picker/date-picker-config.model.ts]: redundant

File: src/app/year-calendar/year-calendar-config.ts (hunk 3 above)
Comment (vlio20) [reply src/app/year-calendar/year-calendar-config.ts]: `multipleYearsNavigateBy` & `showMultipleYearsNavigation`

File: src/app/year-calendar/year-calendar.component.ts (hunk 7 above)
Comment (vlio20) [reply src/app/year-calendar/year-calendar.component.ts]: which one more relevant `EventEmitter<null>`  or EventEmitter<void>? are you emitting null or undefined?
